src/nexusrpc/handler/_common.py

```diff
@@ -0,0 +1,196 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import (
+    Generic,
+    Mapping,
+    Optional,
+)
+
+from nexusrpc.types import OutputT
+
+
+class HandlerErrorType(Enum):
+    """Nexus handler error types.
+
+    See https://github.com/nexus-rpc/api/blob/main/SPEC.md#predefined-handler-errors
+    """
+
+    BAD_REQUEST = "BAD_REQUEST"
+    UNAUTHENTICATED = "UNAUTHENTICATED"
+    UNAUTHORIZED = "UNAUTHORIZED"
+    NOT_FOUND = "NOT_FOUND"
+    RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED"
+    INTERNAL = "INTERNAL"
+    NOT_IMPLEMENTED = "NOT_IMPLEMENTED"
+    UNAVAILABLE = "UNAVAILABLE"
+    UPSTREAM_TIMEOUT = "UPSTREAM_TIMEOUT"
+
+
+class HandlerError(Exception):
+    """
+    A Nexus handler error.
+    """
+
+    __cause__: BaseException
+
+    def __init__(
+        self,
+        message: str,
+        *,
+        type: HandlerErrorType,
+        cause: BaseException,
+        # Whether this error should be considered retryable. If not specified, retry
+        # behavior is determined from the error type. For example, INTERNAL is retryable
+        # by default unless specified otherwise.
+        retryable: Optional[bool] = None,
```

Comment:
In a previous conversation on the TS SDK, it was said that retry behavior should be an enum rather than a boolean, and arguably not optional (though TS will use `undefined` for that).

Those that still applies?

Reply:
I'm not against an enum (vs boolean) if there's an expectation that there might be other values at a later time. I'm only against the use of a named `UNDEFINED` value in that enum in lieu of the `undefined` value.

> I can understand why Python decided to follow TypeScript's lead here

Totally.